### ISSUE TITLE
[ansible_plugins] Handling is broken with collections support

### ISSUE THREAD

**Issue by monsterry** [debops/debops#1320]
Hey @all,

we are currently using debops via git submodule integration and would like to keep it that way. This broke for us with the recent changes to support ansible collections. As for now we were able to distinguish between our roles and the debops roles by setting the `roles_path` accordingly and referencing them via their folder names.

I found that, I could setup a structure simulating a fake collection like that:

```
ansible.cfg
extern
     |
     - ansible-debops (git  submodule ansible-debops.git)
     - collections
         |
         -  ansible_collections
                |
                - debops
                       |
                       - debobs (link to ../../../ansible-debops/ansible/)
```

If I set `collections_paths` in ansible.cfg to  `./extern/collections`, I just have to rename the role imports in our playbooks to use the FQCN (`debops.python` => `debops.debops.python`).

This works for all roles but `debops.ansible_plugins`and all the `debops_contrib.*` roles.

```

The offending line appears to be:
ERROR! the role 'debops.debops.debops-contrib.apparmor' was not found in ...
- import_role:
    name: debops.debops.debops-contrib.apparmor
          ^ here
```
I suggest to rename `debops.ansible_plugins` to `ansible_plugins` or to remove it completely and relay on `plugins` directory on collections level.

If you like release contrib roles in a different collection please separat them into a different directory or name them collections compliant like `contrib_*`.

Of course I could write a shell script linking all roles to their corresponding `debops.*` names and continue with the include via `roles_path`. But I do not like this solution :).

Any help or ideas are appreciated.

**Comment by monsterry** [comment]
I am sorry, I missed that there is an `ansible_plugins` directory containing the same content as `debops.ansible_plugins`. But the problem persists for `debops.contrig*`.

**Comment by drybjed** [comment]
The `debops-contrib.*` roles are currently not included in the DebOps Ansible Collection; they are pending proper integration with the rest of the DebOps roles.

You say that you are using DebOps via `git submodule`, but then you mix Ansible Collection into it. Without full details about how you are using Ansible I can only speculate, but it seems to me that you have you own roles in a Collection and you want to use DebOps roles with them somehow? Could you expand on that a bit? Because at the moment usage via `git submodule` without Ansible Collections mixed in should work as previously - the `debops.` prefix in the roles was dropped to make them work as a Collection, however if you change this in your playbooks/roles, everything should work as previously.

**Comment by monsterry** [comment]
Ok, I will try to elaborate further:

We are using debops via git submodule and are able to manipulate `roles_path` and to reference the debops roles by their name (e.g. `debops.ansible_plugins`). This is totally ok, because we are able to distinguish the origin ofa role by its name and there are no name collisions with our own `ansible_plugins` role.

With your change I am missing the namespacing feature and therefore I tried to simulate a collection setup (without actually using the collection setup) to be able to use the FQCN `debops.debops.ansible_plugins`. Unfortunately this is not possible because the directory is not directly usable as a collection because the conversions from directory to collection, which are done in `make_collection`, are too heavy to compensate them via an outer link structure.

We are also struggling with the use of ansible_galaxy because of the `role1`, `role2`, `role3` crap, which I assume you were forced to use because of the Tarball-limits.

At the moment we are discussing to use an outer link structure to enforce name references like `debops/ansible_plugins`, which is completely incompatible to our former structure.

> The debops-contrib.* roles are currently not included in the DebOps Ansible Collection; they are pending proper integration with the rest of the DebOps roles.

These names are not usable in a collection simulated environment because the FQCN `debops.debops.debops-contrib.*` is simply not recognized correctly ("Role not found").

**Comment by drybjed** [comment]
Looks like you are trying to reap the benefits of Ansible Collections namespacing without actually commiting to Ansible Collections themselves. :-) I had to pass that hurdle as well when converting DebOps to the new structure, it's a hard problem to solve.

Yes, the `roles01`, `roles02`, `roles03` collections were made because of the Ansible Galaxy limits on the tarball size. The shell script that generates the collections tries to move all roles that can be dependencies of other roles to the `debops.debops` collection to keep them "stable", however the placement of roles in a specific collection is not really guaranteed... BTW, what roles are we talking about? Perhaps they could be moved to the `debops.debops` collection as well to keep them "stable".

The contrib roles are indeed a problem. Perhaps in the next development cycle (May-June-July) I'll try to integrate as many of them as possible into proper DebOps roles. Which one of these are you using?

**Comment by monsterry** [comment]
> Looks like you are trying to reap the benefits of Ansible Collections namespacing without actually commiting to Ansible Collections themselves. :-) I had to pass that hurdle as well when converting DebOps to the new structure, it's a hard problem to solve.

Yes exactly, but it was no problem when namespacing was done based on role names.

> Yes, the `roles01`, `roles02`, `roles03` collections were made because of the Ansible Galaxy limits on the tarball size. The shell script that generates the collections tries to move all roles that can be dependencies of other roles to the `debops.debops` collection to keep them "stable", however the placement of roles in a specific collection is not really guaranteed... BTW, what roles are we talking about? Perhaps they could be moved to the `debops.debops` collection as well to keep them "stable".

At the moment we are using:
* environment
* ansible_plugins
* postfix
* postldap
* dovecot
* nginx
* pki
* and and and

Therefore we need a stable setup before we can fully commit to collections. Futhermore the world wide acceptance of collections is not really high and it will be a long time before all role providers will change to that.

> The contrib roles are indeed a problem. Perhaps in the next development cycle (May-June-July) I'll try to integrate as many of them as possible into proper DebOps roles. Which one of these are you using?

We have a test setup using `debops-contrib.dropbear_initramfs`, but it isn't critical to us. I just wanted to mention it.

As mentioned before we tried to simulate the collection by using an outer link structure, but this breaks when it comes to namespaced plugins and so on. I go on and try to update the ansible_plugins role to register fully namespaced plugins per default, so you can skip the conversion in `make_collection`, but it continues to work in non-colllection setups.

**Comment by drybjed** [comment]
At the moment DebOps Collections can be volatile across minor version releases (`v2.x` for example), but they should be stable across patch releases (`v2.0.x`). If you stick to a specific DebOps Collection version, you can use roles in `roles01`, etc. "sub-collection" without worrying that they will suddenly move to a different collection. You can check the [current release schedule](https://docs.debops.org/en/master/news/releases.html#current-release-schedule) to plan accordingly.

Unfortunately the limit on tarball size kind of forces me to split the roles across multiple collections; the mechanism that decides which roles are placed in what collection is kind of intricate, but it tries to move the roles that are used very frequently to the main `debops.debops` collection to keep the naming stable.

I'm not sure what you mean by updating the `ansible_plugins` role... Do you mean changing your own `ansible_plugins` role to be namespaced?

**Comment by monsterry** [comment]
No, my idea was to somehow use the FQCN for plugins in consistent way through debops, even when no collection is used, as this seems to be the ansible preferered way. Therefor ansible_plugins should register the plugins in their FQCN way:

The naive way sadly doesn't work:
```
class FilterModule(object):
    """Register custom filter plugins in Ansible"""

    def filters(self):
        return {
            'parse_kv_config': parse_kv_config,
            'parse_kv_items': parse_kv_items,
            'debops.debops.parse_kv_config': parse_kv_config,
            'debops.debops.parse_kv_items': parse_kv_items
        }
```
I did a little research and it seems like the old plugin style is deprecated (not officially by now) by Ansible and the plugin management changes totally if Ansible detects that collections are used, but you cannot use the legacy way to register FQCN.
